File: docs-examples/example-java/src/test/java/io/micronaut/rabbitmq/docs/consumer/concurrent/ConcurrentSpec.java
package io.micronaut.rabbitmq.docs.consumer.concurrent;

import io.micronaut.context.annotation.Property;
import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
import io.micronaut.test.support.TestPropertyProvider;
import org.junit.jupiter.api.Test;

import java.util.Map;

import static java.util.concurrent.TimeUnit.SECONDS;
import static org.awaitility.Awaitility.await;
import io.micronaut.rabbitmq.testcontainers.RabbitMQ;
import org.junit.jupiter.api.TestInstance;

@MicronautTest
@Property(name = "spec.name", value = "ConcurrentSpec")
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class ConcurrentSpec implements TestPropertyProvider {
    @Override
    public Map<String, String> getProperties() {
        return RabbitMQ.getProperties();
    }

    @Test
    void testConcurrentConsumers(ProductClient productClient, ProductListener productListener) {

        for (int i = 0; i < 4; i++) {
            productClient.send("body".getBytes());
        }


        await().atMost(60, SECONDS).until(() -> productListener.threads.size() == 4);
    }
}

File: docs-examples/example-kotlin/src/test/kotlin/io/micronaut/rabbitmq/docs/consumer/concurrent/ConcurrentSpec.kt
package io.micronaut.rabbitmq.docs.consumer.concurrent

import io.kotest.assertions.nondeterministic.eventually
import io.kotest.core.spec.style.AnnotationSpec
import io.kotest.matchers.shouldBe
import io.micronaut.context.annotation.Property
import io.micronaut.test.extensions.kotest5.annotation.MicronautTest
import io.micronaut.test.support.TestPropertyProvider
import jakarta.inject.Inject
import kotlin.time.Duration.Companion.seconds
import io.micronaut.rabbitmq.testcontainers.RabbitMQ

@MicronautTest
@Property(name = "spec.name", value = "ConcurrentSpec")
class ConcurrentSpec : TestPropertyProvider, AnnotationSpec() {

    override fun getProperties(): Map<String, String> {
        return RabbitMQ.getProperties()
    }

    @Inject
    lateinit var productClient: ProductClient

    @Inject
    lateinit var productListener: ProductListener

    @Test
    suspend fun testBasicProducerAndConsumer() {
        for (i in 0..3) {
            productClient.send("body".toByteArray())
        }

        eventually(5.seconds) {
            productListener.threads.size shouldBe 4
        }
    }
}
